truffleruby/src/main/java/org/truffleruby/core/rubinius/IONodes.java

```diff
@@ -799,7 +801,9 @@ public int accept(DynamicObject io) {
             final int newFd;
 
             try {
-                newFd = ensureSuccessful(nativeSockets().accept(fd, memoryManager().newPointer(address), addressLength));
+                newFd = getContext().getThreadManager().runUntilResult(this,
+                        () -> ensureSuccessful(nativeSockets().accept(fd, memoryManager().newPointer(address), addressLength)),
+                        () -> nativeSockets().shutdown(fd, SHUT_RD));
```

Comment:
It seems `accept` returns `EINVAL` after `shutdown(fd, SHUT_RD);` when trying in a C program.
From the man:
```
 EINVAL Socket is not listening for connections, or addrlen is invalid (e.g., is negative).
```
A quick search seems to show that a socket cannot be reopened after shutdown.

One way to avoid a blocking `accept()` call would be to make it non-blocking with `fcntl(fd, F_SETFL, flags | O_NONBLOCK)`.
Another, probably cleaner as it does not modify the fd flags is to use `select()` with a timeout (like in our IO.select):
```
   In order to be notified of incoming connections on a socket, you can use select(2) or  poll(2).
       A  readable  event  will  be delivered when a new connection is attempted and you may then call
       accept() to get a socket for that connection. 
```